File: docusign-sdk-sample-objc/docusign-sdk-sample-objc/Controllers/TemplatesViewController.m
//
//  TemplatesViewController.m
//  docusign-sdk-sample-objc
//
//  Copyright © 2017 DocuSign. All rights reserved.
//

#import "TemplatesViewController.h"

#import <DocuSignSDK/DocuSignSDK.h>
#import "AppDelegate.h"
#import "DownloadTemplateCell.h"
#import "ProfileManager.h"
#import "ProgressHUD.h"
#import "TemplatesManager.h"
#import "EnvelopesManager.h"

@interface TemplatesViewController ()

@property (nonatomic, strong) TemplatesManager * mTemplatesManager;
@property (nonatomic, strong) EnvelopesManager * mEnvelopesManager;
@property (nonatomic, strong) NSArray<DSMEnvelopeTemplateDefinition *> * mTemplateList;
@property (nonatomic, strong) UIRefreshControl * refreshController;

@end


@implementation TemplatesViewController

static NSString *tableCellId = @"downloadTemplateCell";


#pragma mark - UI Lifecycle Methods

- (void) viewDidLoad
{
    [super viewDidLoad];
    [self styleUIElements];

    // attach template manager
    self.mTemplatesManager = [TemplatesManager sharedInstance];
    self.mEnvelopesManager = [EnvelopesManager sharedInstance];
    
    // add refresh control for table view
    self.refreshController = [[UIRefreshControl alloc] init];
    [self.refreshController addTarget:self action:@selector(handleTableRefresh:)
                     forControlEvents:UIControlEventValueChanged];
    [self.tableView addSubview:self.refreshController];
}


- (void) viewWillAppear:(BOOL)animated
{
    // get list of template definitions
    [self fetchTemplateList];

    // allow landscape orientation for sdk screens
    AppDelegate * appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    appDelegate.restrictRotation = NO;
}

#pragma mark - actions

- (IBAction)onOfflineClaimTapped:(id)sender {
    
    [self promptDevActionWithHandler:^(DSMSigningMode signingMode) {
        if (@available(iOS 11.0, *)) {
            [self.mEnvelopesManager presentComposeEnvelopeViewControllerWithPresentingController:self signingMode:signingMode];
        } else {
            UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"iCloud Entitlement required" message:@"For iOS 10 and below, iCloud entitlements must be added and DSM_SETUP_ICLOUD_DOCUMENT_ENABLED set to true for document picker usage." preferredStyle:UIAlertControllerStyleAlert];
            UIAlertAction *action = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                [self.mEnvelopesManager presentComposeEnvelopeViewControllerWithPresentingController:self signingMode:signingMode];
            }];
            
            [alertController addAction:action];
            [self presentViewController:alertController animated:YES completion:nil];
        }
    }];
    
}

#pragma mark - TableView Delegate Methods

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [self.mTemplateList count];
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    // reuse cell object (or create a new object if needed)
    DownloadTemplateCell *cell = [tableView dequeueReusableCellWithIdentifier:tableCellId];
    if (cell == nil)
    {
        [tableView registerNib:[UINib nibWithNibName:@"DownloadTemplateCell" bundle:nil] forCellReuseIdentifier:tableCellId];
        cell = [tableView dequeueReusableCellWithIdentifier:tableCellId];
    }
    return cell;
}


- (void)tableView:(UITableView *)tableView willDisplayCell:(nonnull DownloadTemplateCell *)cell forRowAtIndexPath:(nonnull NSIndexPath *)indexPath
{
    DSMEnvelopeTemplateDefinition * template = [self.mTemplateList objectAtIndex:indexPath.row];

    // set name label
    cell.nameLabel.text = template.name;
    
    // determine image for accessory button
    if ([self.mTemplatesManager templateIsCachedWithId:template.templateId]) {
        // template is already cached
        [cell setAccessoryType:UITableViewCellAccessoryDisclosureIndicator];
        [cell.downloadButton setBackgroundImage:nil forState:UIControlStateNormal];
    }
    else
    {
        // template needs to be downloaded
        [cell.downloadButton setBackgroundImage:[UIImage imageNamed:@"download"] forState:UIControlStateNormal];
        [cell.downloadButton addTarget:self action:@selector(downloadTemplate:withEvent:)
                      forControlEvents:UIControlEventTouchUpInside];
    }
}


- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    DSMEnvelopeTemplateDefinition * template = [self.mTemplateList objectAtIndex:indexPath.row];
    
    // if template is already cached, then move to policy screen
    if ([self.mTemplatesManager templateIsCachedWithId:template.templateId])
    {
        // capture the selected template id
        [[ProfileManager sharedInstance] setClientTemplateId:template.templateId];
        
        // segue to policy screen
        [self performSegueWithIdentifier:@"segue_claim_policy"
                                  sender:[self.tableView cellForRowAtIndexPath:indexPath]];
    }
}


#pragma mark - Private Methods

- (void) fetchTemplateList
{
    [self.mTemplatesManager getTemplateListWithCompletion:^(NSArray<DSMEnvelopeTemplateDefinition*>* templates) {
        
        // store template defitions
        self.mTemplateList = templates;
        
        // refresh table view
        [self.tableView reloadData];
    }];
}


- (void) downloadTemplate:(UIControl *)button withEvent:(UIEvent *)event
{
    long rowIndex = [self getIndexForTableCell:event];
    
    [ProgressHUD show:@"Downloading..."];
    
    NSString * templateId = [self.mTemplateList objectAtIndex:rowIndex].templateId;
    [self.mTemplatesManager cacheTemplateWithId:templateId withCompletion:^(NSString * errorMsg) {
        [ProgressHUD dismiss];
        
        if (errorMsg == nil)
        {
            [self.tableView reloadData];
        }
        else
        {
            [self displayErrorPrompt:errorMsg];
        }
    }];
}


- (void)handleTableRefresh: (id)sender
{
    // retrieve template list
    [self fetchTemplateList];
    [self.refreshController endRefreshing];
}


- (long) getIndexForTableCell:(UIEvent *)event
{
    // determine the index of item tapped
    NSSet *touches = [event allTouches];
    UITouch *touch = [touches anyObject];
    CGPoint currentTouchPosition = [touch locationInView:self.tableView];
    NSIndexPath *indexPath = [self.tableView indexPathForRowAtPoint: currentTouchPosition];
    
    if (indexPath == nil)
        return -1;
    else
        return indexPath.row;
}


- (void) displayErrorPrompt:(NSString *)errorMessage
{
    NSLog(@"Error: %@", errorMessage);
    
    // display error alert
    UIAlertController * alertController = [UIAlertController alertControllerWithTitle:@"Error" message:errorMessage
                                                                       preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction * okAction = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleDefault handler:nil];
    [alertController addAction:okAction];
    [self presentViewController:alertController animated:YES completion:nil];
}


- (void) styleUIElements
{
    // add custom nav title
    UIView * navTitleView = [[[NSBundle mainBundle] loadNibNamed:@"CustomNavTitle" owner:self options:nil] lastObject];
    self.navigationItem.titleView = navTitleView;
}

- (void) promptDevActionWithHandler:(void(^)(DSMSigningMode signingMode))handler
{
    UIAlertAction *onlineAction = [UIAlertAction actionWithTitle:@"Online Envelope" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        handler(DSMSigningModeOnline);
    }];
    UIAlertAction *offlineAction = [UIAlertAction actionWithTitle:@"Offline Envelope" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        handler(DSMSigningModeOffline);
    }];
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil];
    
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Developer's Notes" message:@"You can either compose an envelope in online or offline mode. You would need to check for network connectivity and present the appropriate view controller." preferredStyle:UIAlertControllerStyleAlert];
    
    [alertController addAction:onlineAction];
    [alertController addAction:offlineAction];
    [alertController addAction:cancelAction];
    [self presentViewController:alertController animated:YES completion:nil];
}

@end

File: docusign-sdk-sample-swift/docusign-sdk-sample-swift/Controllers/TemplatesViewController.swift
//
//  TemplatesViewController.swift
//  docusign-sdk-sample-swift
//
//  
//  Copyright © 2017 DocuSign. All rights reserved.
//

import DocuSignSDK
import SVProgressHUD
import UIKit


class TemplatesViewController: UIViewController, UITableViewDelegate, UITableViewDataSource, ContainerWithTableView
{
    // MARK: UI Elements

    @IBOutlet var tableView: UITableView!

    // MARK: Private Variables / Members

    private let cellReuseIdentifier = "cell_template";
    private var mTemplateList = [DSMEnvelopeTemplateDefinition]();
    private var refreshControl: UIRefreshControl!


    // MARK: UI Lifecycle Methods
    
    override func viewDidLoad()
    {
        super.viewDidLoad()

        tableView.delegate = self;
        tableView.dataSource = self;
        
        let nib = UINib(nibName: "TemplateTableViewCell", bundle: nil)
        tableView.register(nib, forCellReuseIdentifier: cellReuseIdentifier)
        
        // add refresh control to table view
        refreshControl = UIRefreshControl()
        refreshControl.attributedTitle = NSAttributedString(string: "Pull to refresh")
        refreshControl.addTarget(self, action: #selector(refresh), for: UIControlEvents.valueChanged);
        tableView.addSubview(refreshControl)
        
        // fetch list of template definitions
        TemplatesManager.sharedInstance.getTemplateListWithCompletion { (templateDefinitions) in
            if (templateDefinitions != nil)
            {
                self.mTemplateList = templateDefinitions!;
                self.tableView.reloadData();
            }
        }
    }
    
    // MARK: TableView Methods
    
    // number of sections in table view
    func numberOfSections(in tableView: UITableView) -> Int
    {
        return 1
    }
    
    
    // number of rows in table view
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        return mTemplateList.count;
    }
    
    
    // create a cell for each table view row
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell
    {
        // create a new cell if needed or reuse an old one
        let cell: TemplateTableViewCell =  self.tableView.dequeueReusableCell(withIdentifier: cellReuseIdentifier) as! TemplateTableViewCell!
        
        return cell;
    }


    // configure cell before display
    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath)
    {
        let template: DSMEnvelopeTemplateDefinition = mTemplateList[indexPath.row];
        let templateCell: TemplateTableViewCell = cell as! TemplateTableViewCell;
        templateCell.lbl_templateName.text = template.name;
        
        // set image and action for download button
        if (TemplatesManager.sharedInstance.templateIsCachedWithId(templateId: template.templateId))
        {
            // template is already cached
            templateCell.btn_download.setBackgroundImage(UIImage(named: "download_done"), for: UIControlState.normal);
            templateCell.btn_download.removeTarget(self, action: nil, for: .touchUpInside);
        }
        else
        {
            // template needs to be downloaded
            templateCell.btn_download.setBackgroundImage(UIImage(named: "download"), for: UIControlState.normal);
            templateCell.btn_download.addTarget(self, action: #selector(downloadButtonTapped(_:)), for: .touchUpInside)
        }
    }
    

    func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath)
    {
        if (editingStyle == UITableViewCellEditingStyle.delete)
        {
            // remove specified template
            let templateId = self.mTemplateList[indexPath.row].templateId;
            TemplatesManager.sharedInstance.removeTemplateWithId(templateId: templateId!);
            
            // reload table to update template download status
            self.tableView.reloadData();
        }
    }


    func tableView(_ tableView: UITableView, titleForDeleteConfirmationButtonForRowAt indexPath: IndexPath) -> String? {
        return "Uncache";
    }
    
    
    // MARK: ContainerWithTableView Protocol Methods
    
    func startEditModeOnTableView()
    {
        tableView.setEditing(true, animated: true);
    }
    
    
    func endEditModeOnTableView()
    {
        tableView.setEditing(false, animated: true);
    }


    // MARK: Selector or Helper Methods

    func downloadButtonTapped(_ sender:UIButton)
    {
        if let cell = sender.superview?.superview as? UITableViewCell
        {
            SVProgressHUD.show(withStatus: "Downloading...");
            
            // determine table row index and associated template id
            let indexPath = self.tableView.indexPath(for: cell)
            let templateId = mTemplateList[(indexPath?.row)!].templateId;
        
            if (templateId != nil) {
                // cache the specified template
                TemplatesManager.sharedInstance.cacheTemplateWithId(templateId: templateId!) { (errMsg) in
                    SVProgressHUD.dismiss();
                    
                    if (errMsg != nil)
                    {
                        self.displayErrorPrompt(errMsg: errMsg!);
                    }
                    else
                    {
                        self.tableView.reloadData();
                    }
                }
            }
        }
    }


    // MARK: Private Methods
    
    func refresh(_ sender:Any) {

        TemplatesManager.sharedInstance.getTemplateListWithCompletion { (templateDefinitions) in
            if (templateDefinitions != nil)
            {
                self.mTemplateList = templateDefinitions!;
                self.tableView.reloadData();
                self.refreshControl.endRefreshing()
            }
        }
    }

    
    private func displayErrorPrompt(errMsg: String)
    {
        NSLog("Display error prompt: " + errMsg);
        
        let alert = UIAlertController(title: "Error", message: errMsg, preferredStyle: UIAlertControllerStyle.alert)
        alert.addAction(UIAlertAction(title: "Ok", style: UIAlertActionStyle.default, handler: nil))
        self.present(alert, animated: true, completion: nil)
    }

}
